### PR TITLE
[WIP] builder/vmware-esxi: remove floppy files when done

### LINKED ISSUE #4679: [question] vmware-iso remote esxi packer_cache cleanup
Packer Version: 0.12.3
Builder: vmware-iso / esxi

After uploading the ISO media to the ESXI Host it seems that it does not clean up the temporal files created during the upload.

I wonder if there is a way to explicitly specify this or maybe should be considered a missing step after the checksum verification.

```shell
packer_cache # ls -alh
total 9188352
drwxr-xr-x    1 root     root        9.0K Mar 17 16:45 .
drwxr-xr-t    1 root     root        1.4K Mar 17 19:14 ..
-rw-r--r--    1 root     root        5.0G Mar 13 14:20 50d67a08a88bc6f7b7dff4779c266c3f898c5cf1b11151e782c4505ecdcc5727.iso
-rw-r--r--    1 root     root        3.6G Feb 27 16:32 e5a35a4e2b4a6e93f848a6c0ec4c9620154fdb8f0166895e3b25ed710b44916c.iso
-rw-r--r--    1 root     root        1.4M Mar 14 14:46 packer043878966
-rw-r--r--    1 root     root        1.4M Mar 15 16:01 packer066039411
-rw-r--r--    1 root     root        1.4M Mar  2 00:10 packer080882081
-rw-r--r--    1 root     root        1.4M Mar  1 19:33 packer082312476
-rw-r--r--    1 root     root        1.4M Mar  2 00:42 packer111817382
-rw-r--r--    1 root     root        1.4M Mar 13 23:25 packer119951066
-rw-r--r--    1 root     root        1.4M Mar  1 23:57 packer142557212
-rw-r--r--    1 root     root        1.4M Mar 13 14:09 packer144182597
-rw-r--r--    1 root     root        1.4M Mar  2 00:20 packer144286955
-rw-r--r--    1 root     root        1.4M Mar 10 16:22 packer154908027
-rw-r--r--    1 root     root        1.4M Mar  1 23:58 packer161618364
-rw-r--r--    1 root     root        1.4M Mar  1 22:15 packer176406123
-rw-r--r--    1 root     root        1.4M Mar 14 19:32 packer189899588
-rw-r--r--    1 root     root        1.4M Mar  2 00:45 packer214826305
-rw-r--r--    1 root     root        1.4M Mar  1 18:40 packer238955964
-rw-r--r--    1 root     root        1.4M Mar 13 19:30 packer253631037
-rw-r--r--    1 root     root        1.4M Mar 13 14:52 packer265375062
-rw-r--r--    1 root     root        1.4M Mar 13 14:43 packer283137941
-rw-r--r--    1 root     root        1.4M Mar 15 19:30 packer315126782
-rw-r--r--    1 root     root        1.4M Mar 13 14:26 packer380691963
-rw-r--r--    1 root     root        1.4M Mar 14 21:38 packer388087604
-rw-r--r--    1 root     root        1.4M Mar  1 18:55 packer389555545
-rw-r--r--    1 root     root        1.4M Mar 10 16:20 packer393754863
-rw-r--r--    1 root     root        1.4M Mar  7 22:00 packer413680810
-rw-r--r--    1 root     root        1.4M Mar  1 19:26 packer422342266
-rw-r--r--    1 root     root        1.4M Mar 10 21:32 packer437742730
-rw-r--r--    1 root     root        1.4M Feb 27 22:59 packer528255518
-rw-r--r--    1 root     root        1.4M Mar 10 21:35 packer531381278
-rw-r--r--    1 root     root        1.4M Mar  2 00:03 packer547141835
-rw-r--r--    1 root     root        1.4M Mar  6 18:08 packer553108740
-rw-r--r--    1 root     root        1.4M Mar  1 18:27 packer556331357
-rw-r--r--    1 root     root        1.4M Mar  7 19:03 packer556780646
-rw-r--r--    1 root     root        1.4M Mar  2 00:17 packer583660028
-rw-r--r--    1 root     root        1.4M Feb 27 23:07 packer586694884
-rw-r--r--    1 root     root        1.4M Mar  6 14:44 packer604715198
-rw-r--r--    1 root     root        1.4M Mar  1 22:08 packer606695494
-rw-r--r--    1 root     root        1.4M Mar  2 00:07 packer625115238
-rw-r--r--    1 root     root        1.4M Mar  1 19:40 packer630227181
-rw-r--r--    1 root     root        1.4M Mar  1 21:56 packer630566133
-rw-r--r--    1 root     root        1.4M Mar 13 15:38 packer632782751
-rw-r--r--    1 root     root        1.4M Mar  7 15:57 packer643845539
-rw-r--r--    1 root     root        1.4M Mar 10 15:28 packer651778325
-rw-r--r--    1 root     root        1.4M Mar  1 18:45 packer653411759
-rw-r--r--    1 root     root        1.4M Mar  2 00:38 packer671712010
-rw-r--r--    1 root     root        1.4M Mar 13 22:54 packer742391307
-rw-r--r--    1 root     root        1.4M Mar  1 19:07 packer753142852
-rw-r--r--    1 root     root        1.4M Mar 14 18:51 packer763740951
-rw-r--r--    1 root     root        1.4M Mar  7 19:07 packer795197586
-rw-r--r--    1 root     root        1.4M Mar 15 00:27 packer805983442
-rw-r--r--    1 root     root        1.4M Feb 28 02:11 packer812049395
-rw-r--r--    1 root     root        1.4M Mar 13 14:59 packer818549044
-rw-r--r--    1 root     root        1.4M Mar  1 23:59 packer819451376
-rw-r--r--    1 root     root        1.4M Mar  7 15:35 packer872332728
-rw-r--r--    1 root     root        1.4M Mar 13 18:26 packer885228962
-rw-r--r--    1 root     root        1.4M Mar 10 20:23 packer886819084
-rw-r--r--    1 root     root        1.4M Mar  7 19:19 packer891107791
-rw-r--r--    1 root     root        1.4M Mar 14 20:50 packer891995528
-rw-r--r--    1 root     root        1.4M Feb 28 15:39 packer895083033
-rw-r--r--    1 root     root        1.4M Mar 10 21:38 packer907125088
-rw-r--r--    1 root     root        1.4M Mar 14 20:04 packer930686243
-rw-r--r--    1 root     root        1.4M Mar 17 17:14 packer953889151
-rw-r--r--    1 root     root        1.4M Mar  2 00:31 packer967793180
```

### PR DESCRIPTION
**Pull request by mwhooker**
need to test

Closes #4679

### CODE REVIEW

**Comment by lmayorga1980** [comment]
```shell
 ~/go/src/github.com/mitchellh/packer/bin/packer build -only=vmware-iso -var-file=rhel68-variables-home.json rhel68.json 
vmware-iso output will be in this color.

Warnings for build 'vmware-iso':

* A checksum type of 'none' was specified. Since ISO files are so big,
a checksum is highly recommended.

==> vmware-iso: Downloading or copying ISO
    vmware-iso: Downloading or copying: file:///Users/lmo0/git/packer-templates/iso/rhel-server-6.8-x86_64-dvd.iso
==> vmware-iso: Creating floppy disk...
    vmware-iso: Copying files flatly from floppy_files
    vmware-iso: Copying file: http/cfg_redhat/ks.cfg
    vmware-iso: Done copying files from floppy_files
    vmware-iso: Collecting paths from floppy_dirs
    vmware-iso: Resulting paths from floppy_dirs : []
    vmware-iso: Done copying paths from floppy_dirs
==> vmware-iso: Uploading Floppy to remote machine...
==> vmware-iso: Uploading ISO to remote machine...
==> vmware-iso: Creating virtual machine disk
==> vmware-iso: Building and writing VMX file
==> vmware-iso: Registering remote VM...
==> vmware-iso: Starting virtual machine...
    vmware-iso: The VM will be run headless, without a GUI. If you want to
    vmware-iso: view the screen of the VM, connect via VNC with the password "qTD5ODpx" to
    vmware-iso: vnc://192.168.1.25:5900
==> vmware-iso: Waiting 10s for boot...
==> vmware-iso: Connecting to VM via VNC
==> vmware-iso: Typing the boot command over VNC...
==> vmware-iso: Waiting for SSH to become available...
==> vmware-iso: Connected to SSH!
==> vmware-iso: Provisioning with shell script: scripts/rhn_reg.sh
    vmware-iso: Registering to: subscription.rhn.redhat.com:443/subscription
    vmware-iso: The system has been registered with ID: fa1e66ff-467a-4338-9e1c-923f418b6896
    vmware-iso:
    vmware-iso: Installed Product Current Status:
    vmware-iso: Product Name: Red Hat Enterprise Linux Server
    vmware-iso: Status:       Subscribed
    vmware-iso:
    vmware-iso: Repository 'rhel-6-server-extras-rpms' is enabled for this system.
    vmware-iso: Repository 'rhel-6-server-optional-rpms' is enabled for this system.
==> vmware-iso: Provisioning with shell script: scripts/base.sh
==> vmware-iso: Gracefully halting virtual machine...
    vmware-iso: Waiting for VMware to clean up after itself...
==> vmware-iso: Deleting unnecessary VMware files...
    vmware-iso: Deleting: /vmfs/volumes/datastore1/<MYHOST>/vmware.log
==> vmware-iso: Compacting the disk image
==> vmware-iso: Cleaning VMX prior to finishing up...
    vmware-iso: Unmounting floppy from VMX...
    vmware-iso: Detaching ISO from CD-ROM device...
    vmware-iso: Disabling VNC server...
==> vmware-iso: Keeping virtual machine registered with ESX host (keep_registered = true)
Build 'vmware-iso' finished.
```
From the ESXI Host

```
packer_cache # ls -alh
total 3790856
drwxr-xr-x    1 root     root         560 Mar 20 00:22 .
drwxr-xr-t    1 root     root        1.5K Mar 20 00:17 ..
-rw-r--r--    1 root     root        1.4M Mar 19 23:54 packer617471951
-rw-r--r--    1 root     root        3.6G Mar 19 23:57 rhel-server-6.8-x86_64-dvd.iso
```
Note: I was having issues with my personal ESXI so ... lots of Ctrl+C (around 10+)

**Comment by mwhooker** [comment]
hmm, I wonder if this delete is being called at the wrong time. It should probably happen after floppy unmount. Would be handy to see the log output when run with `PACKER_LOG=1`, though.

**Comment by lmayorga1980** [comment]
Well,

Running a different test now with empty */packer_cache*. My take is that sometimes when I type Ctrl+C it happens to hang and then had to Ctrl+Z and delete my VM files manually. Some sort of packer<#$%(%> orphan files?.  

Now the floppy files were cleaned successfully.

**Comment by rismoney** [comment]
any chance that this can get some attention?  noticed my iso_cache had 1000s of packer files in it...

**Comment by mwhooker** [comment]
@rismoney if you can confirm this works for you I'll merge it

**Comment by GMZwinge** [comment]
@mwhooker We are seeing this issue and I can help with the testing.

**Comment by mwhooker** [comment]
@GMZwinge here ya go. I also changed the code a bit so hopefully it will actually work. [packer-4682-windows-amd64.zip](https://github.com/hashicorp/packer/files/1952388/packer-4682-windows-amd64.zip)

**Comment by GMZwinge** [comment]
@mwhooker According to the log, it seems to remove the floppy file in the ESXi. But it also removes the whole packer_cache folder on the ESXi. This is a problem. This folder caches OS iso files so that they don't need to be copied every time. In the log, it's this line that is the problem: 2018/04/26 14:23:38 packer.exe: 2018/04/26 14:23:38 [DEBUG] starting remote command: rm -rf /vmfs/volumes/datastore2/packer_cache. I cannot confirm that the floppy file itself is indeed removed because the whole packer_cache folder that contains it is removed.

**Comment by GMZwinge** [comment]
It looks like it working. But the log shows that it's running this command: "2018/04/26 15:10:03 packer.exe: 2018/04/26 15:10:03 [DEBUG] starting remote command: rm -rf /vmfs/volumes/datastore2/packer_cache/packer703904011". I would change the "rm -rf " into "rm -f ". It is removing a single file and doesn't need to be recursive and safer in case for some reason the floppy file name is empty.

File: builder/vmware/iso/step_remote_upload.go [hunk 1]
```diff
@@ -49,4 +50,27 @@ func (s *stepRemoteUpload) Run(state multistep.StateBag) multistep.StepAction {
...
+		return
+	}
+
+	path, ok := state.Get(s.Key).(string)
+	if !ok {
+		return
+	}
+
+	log.Printf("Cleaning up remote path: %s", path)
+	err := remote.remove(path)
```
Comment (GMZwinge) [inline builder/vmware/iso/step_remote_upload.go]: Could it use remote.Remove(path)?

Comment (mwhooker) [reply builder/vmware/iso/step_remote_upload.go]: this PR is outdated and I'll need to open a new one

**Comment by GMZwinge** [comment]
Thank you @mwhooker. Anything else I can do?

**Comment by mwhooker** [comment]
@GMZwinge thanks for all the help testing. If you wanted to go to https://github.com/hashicorp/packer/pull/6206 and +1 the code changes, I think we can merge

**Comment by GMZwinge** [comment]
The only thing I could +1 in #6206 is your first comment. I did not see how I could +1 anything under the tab "Files changed". May be I missed it, or I don't have the privilege.

**Pull request closed without merging**